### PR TITLE
Atomically check whether inplace instance upgrade is in progress

### LINKED ISSUE #654: instance manager's /update insufficient check to prevent parallel upgrades
HTTP handler for `/update` checks and sets `InstanceManagerIsUpgrading` non atomically, which makes it possible for more than one parallel upgrade processes to proceed

### PR DESCRIPTION
**Pull request by redbaron**
Fixes #654

### CODE REVIEW

File: pkg/management/upgrade/upgrade.go [hunk 1]
```diff
@@ -53,12 +53,10 @@ func FromReader(
 	instance *postgres.Instance,
 	r io.Reader,
 ) error {
-	// Mark this instance manager as upgrading
-	instance.InstanceManagerIsUpgrading = true
 	defer func() {
 		// Something didn't work if we reach this point, as this process
 		// should have already been replaced by a new one with the new binary
-		instance.InstanceManagerIsUpgrading = false
+		instance.InstanceManagerIsUpgrading.Store(false)
```
Comment (phisco) [inline pkg/management/upgrade/upgrade.go:59]: Given that now we are setting this to true outside this method, I'd move the reset too. WDYT?

**Comment by mnencia** [comment]
I prefer to add a comment explaining why we do not need to store `false` if there is no error during the upgrade (that function never returns on success).

**Comment by leonardoce** [comment]
Well, we could improve the old comment:

```
		// Unfortunately this point, if everything is right, will not be reached.
		// At this stage we are running the new version of the instance manager
		// and not the old one.
```

File: pkg/management/postgres/webserver/remote.go [hunk 2]
```diff
@@ -142,14 +142,15 @@ func (ws *remoteWebserverEndpoints) updateInstanceManager(
 		}
 
 		// No need to do anything if we are already upgrading
-		if ws.instance.InstanceManagerIsUpgrading {
+		if !ws.instance.InstanceManagerIsUpgrading.CompareAndSwap(false, true) {
 			http.Error(w, "instance manager is already upgrading", http.StatusTeapot)
```
Comment (redbaron) [inline pkg/management/postgres/webserver/remote.go:146]: I just noticed `StatusTeapot` :) Maybe 409 Conflict is more appropriate?

**Comment by leonardoce** [comment]
I removed the code implementing the temporary file name, and put it in #659 to make review easier.

File: pkg/management/postgres/webserver/remote.go (hunk 2 above)
Comment (leonardoce) [reply pkg/management/postgres/webserver/remote.go:146]: Yeah, I don't have strong opinions on that!

**Review by leonardoce — APPROVED** [review]

**Review by jsilvela — APPROVED** [review]

**Comment by jsilvela** [comment]
The logic for unsetting the `InstanceManagerIsUpgrading` was a bit hidden and magical.
I have made an explicit `defer` for the unsetting.

**Comment by litaocdl** [comment]
The latest e2e triggered is https://github.com/cloudnative-pg/cloudnative-pg/actions/runs/2968471359, not sure why the unit test statuses could not be updated.

**Pull request merged**